Add MaxNoncurrentVersions to NoncurrentExpiration action

krisis opened the pull request:
## Description
This unit allows users to limit the maximum number of noncurrent versions of an
object.


## Motivation and Context
To limit number of noncurrent versions of an object.

## How to test this PR?
```shell
#!/bin/bash

ALIAS=$1

./mc mb $ALIAS/mybucket
./mc version enable $ALIAS/mybucket
for i in {1..10}
do
   ./mc cp /etc/issue $ALIAS/mybucket/obj-1
done
cat <<EOF | ./mc ilm import $ALIAS/mybucket
{
    "Rules": [
        {
            "ID": "test-max-noncurrent",
            "Status": "Enabled",
            "Filter": {},
            "NoncurrentVersionExpiration": {
                "MaxNoncurrentVersions": 5
            }
        }
    ]
}
EOF

```
`bash -x <above-script.sh> myminio`

Now, perform `mc cp` to upload a new version and trigger the removal of excess noncurrent versions as configured.

This requires mc compiled with https://github.com/minio/minio-go/pull/1580

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Documentation updated
- [ ] Unit tests added/updated

vadmeste commented on internal/bucket/lifecycle/lifecycle.go:
```diff
@@ -468,3 +472,18 @@ func (lc Lifecycle) TransitionTier(obj ObjectOpts) string {
+		if rule.NoncurrentVersionExpiration.MaxNoncurrentVersions == 0 {
+			continue
+		}
+		if lim == 0 || lim > rule.NoncurrentVersionExpiration.MaxNoncurrentVersions {
```
I think you need to return the lowest max versions limit from all rules, not the highest one

krisis replied:
This method is returning the lowest max versions limit already. I shall add a unit test that verifies this.

vadmeste commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +266,65 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+func ilmLimitNoncurrentVersions(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (err error) {
+	var lc *lifecycle.Lifecycle
+	if lc, err = globalLifecycleSys.Get(bucket); err != nil {
+		if errors.Is(err, BucketLifecycleNotFound{Bucket: bucket}) {
```
You don't have to set bucket name here

krisis replied:
I think it would be useful to know for which bucket this function failed in, when it does.

harshavardhana replied:
Yeah but `Go` will be only doing `type` check here not the Struct value @krisis

krisis replied:
`errors.Is` compares for equality if the value is comparable. ref: https://play.golang.org/p/YHYiA0-Ox7f

vadmeste commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +266,65 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+		if errors.Is(err, BucketLifecycleNotFound{Bucket: bucket}) {
+			return nil
+		}
+		return err
```
Maybe we need log the error and return nil instead.

krisis replied:
Are you suggesting that we log the "Bucket not found" error too? All errors returned by this function are logged by the caller.

vadmeste reviewed (commented):
Some comments

poornas commented on internal/bucket/lifecycle/expiration.go:
```diff
@@ -23,11 +23,12 @@ import (
+	errLifecycleInvalidExpiration           = Errorf("Exactly one of Days (positive integer) or Date (positive ISO 8601 format) should be present inside Expiration.")
+	errLifecycleInvalidDeleteMarker         = Errorf("Delete marker cannot be specified with Days or Date in a Lifecycle Expiration Policy")
+	errLifecycleDateNotMidnight             = Errorf("'Date' must be at midnight GMT")
+	errLifecycleInvalidNoncurrentExpiration = Errorf("Exactly one of NoncurrentDays (positive integer) or NoncurrentVersions should be specified an NoncurrentExpiration rule.")
```
```suggestion
	errLifecycleInvalidNoncurrentExpiration = Errorf("Exactly one of NoncurrentDays (positive integer) or NoncurrentVersions should be specified in a NoncurrentExpiration rule.")
```

poornas commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +266,65 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+			break
+		}
+	}
+
```
if len(noncurrentVersions) == 0 {
              return nil
        }

klauspost commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +266,69 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+		keep               int
+	)
+	for {
+		info, err = objectAPI.ListObjectVersions(ctx, bucket, object, info.NextMarker, info.NextVersionIDMarker, "", 1000)
```
No. This is way too expensive as an ILM call. You must find another way for this, otherwise you will tank performance. Callers should have enough information to decide if we need to truncate versions before they even make the call.

If you cannot decide this in the caller it is better NOT to have the call.

For scanner, we have all versions [here](https://github.com/minio/minio/blob/bb639d9f29fdd7595a1b520d8ea533c83acaa607/cmd/xl-storage.go#L458) - although only from one disk. Use that as a smoke screen test, and only do `ListObjectVersions` (or preferably a cheaper call) when you have seen more than the number of versions you want.

In terms of structure, add a second callback similar to `item.applyActions(...)`, that sends all versions and return a possibly modified list of versions.

Callback could be something like:
```Go
		fivs, err = item.applyVersionActions(ctx, fivs)
		if err != nil {
			if intDataUpdateTracker.debug {
				console.Debugf(color.Green("scannerBucket:")+" applying version actions failed: %v: %w\n", item.Path, err)
			}
			return sizeSummary{}, errSkipFile
		}

```

harshavardhana replied:
This is only considered as okay solution @klauspost but not an optimal one.

Since we have to revisit this area subsequently once #13573

klauspost replied:
@harshavardhana It will be *very* slow. We should not do a release with this IMO.

harshavardhana replied:
As per @krisis testing, it isn't since its per object @klauspost - perhaps run `warp` and lets see how much we take a hit with and without it. 

NOTE: we should be optimizing here to make sure do GetObjectInfo() since its per object.

krisis replied:
@klauspost this function is called in `PutObjectHandler`, `CopyObjectHandler` and `CompleteMultipartUploadHandler`. It is not called by the scanner. It is only ever called for objects uploaded to a bucket with `MaxNoncurrentVersions` configured.
The warp tests I ran locally were to compare the performance difference between master (then) and this PR, *without* `MaxNoncurrentVersions` configured. This was to learn if this feature affects performance when disabled. I didn't find any noticeable difference.

Initially, I set out to implement this by adding a DeleteNoncurrentVersions method to the `ObjectLayer` and `StorageAPI` interface in order to push down the `MaxNoncurrentVersions` value. This way the listing/deletion happens at the participating disks. @harshavardhana recommended the current approach knowing that we will re-implement this taking advantage of https://github.com/minio/minio/pull/13573.

> If you cannot decide this in the caller it is better NOT to have the call.

With the background out of the way, I am open to the idea of moving application of this ILM rule exclusively to the scanner. At the scanner, we can do as you suggest. I believe that the scanner will be able to catch up on those objects whose versions are growing at a very fast rate, to remove excess noncurrent versions in batches. Does this sound good to you?

klauspost commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +266,69 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+	)
+	for {
+		info, err = objectAPI.ListObjectVersions(ctx, bucket, object, info.NextMarker, info.NextVersionIDMarker, "", 1000)
+		if err != nil {
```
I can add a function that will give you cheap quorum summary of versions in #13573 - could also be good for healing.

harshavardhana commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +267,73 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+func ilmLimitNoncurrentVersions(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (err error) {
+	if enable := env.Get("_MINIO_LIMIT_NONCURRENT", ""); enable == "" {
+		return nil
+	}
```
```suggestion
	if env.Get("MINIO_API_ILM_INLINE", config.EnableOff) == config.EnableOff {
		return nil
	}
```

harshavardhana commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +271,100 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+			Versioned:        versioned,
+			VersionSuspended: versionSuspended,
+		})
+		for _, dobj := range deletedObjs {
```
We only notify upon success @krisis - this is notifying blindly for all objects.

krisis replied:
done

harshavardhana commented on cmd/data-scanner.go:
```diff
@@ -970,13 +971,61 @@ func (i *scannerItem) applyTierObjSweep(ctx context.Context, o ObjectLayer, oi O
+		})
+	}
+
+	go deleteObjectVersions(GlobalContext, o, i.bucket, toDel)
```
There is no limit to the number of go-routines this can create - are you okay with this? @krisis

harshavardhana commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +272,104 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+			if locked {
+				continue
+			}
+		}
```
```suggestion
		if rcfg.LockEnabled && enforceRetentionForDeletion(ctx, oi) {
			continue
		}
```

should be simpler?

harshavardhana commented on cmd/object-handlers-common.go:
```diff
@@ -264,3 +272,104 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
+			}
+			// log the first error
+			logger.LogIf(ctx, err)
+			return
```
should we proceed to delete the rest here? instead of `return` ?

krisis committed:
ilm: Add support for MaxNoncurrentVersions

This unit allows users to limit the maximum number of noncurrent versions of an
object.

To enable this rule,
```
cat <<EOF | ./mc ilm import myminio/mybucket
{
    "Rules": [
        {
            "ID": "test-max-noncurrent",
            "Status": "Enabled",
            "Filter": {
                "Prefix": "user-uploads/"
            },
            "NoncurrentVersionExpiration": {
                "MaxNoncurrentVersions": 5
            }
        }
    ]
}
EOF

krisis committed:
Combine error logging and notifications

krisis committed:
Fix lint errors

harshavardhana committed:
remove conflict resolution issues

klauspost commented on cmd/data-scanner.go:
```diff
@@ -970,13 +971,57 @@ func (i *scannerItem) applyTierObjSweep(ctx context.Context, o ObjectLayer, oi O
+
+// applyVersionActions will apply lifecycle checks on all versions of a scanned item. Returns versions that remain
+// after applying lifecycle checks configured.
+func (i *scannerItem) applyVersionActions(ctx context.Context, o ObjectLayer, fivs []FileInfo, sizeS *sizeSummary) ([]FileInfo, error) {
```
I think we can omit `sizeSummary` here. We parse each version anyway for that.

klauspost commented on cmd/storage-datatypes.go:
```diff
@@ -93,7 +93,8 @@ type FileInfoVersions struct {
 
-	Versions []FileInfo `msg:"vs"`
+	Versions     []FileInfo `msg:"vs"`
+	FreeVersions []FileInfo `msg:"fvs"`
```
API bump needed.

harshavardhana commented on cmd/api-errors.go:
```diff
@@ -577,6 +578,11 @@ var errorCodes = errorCodeMap{
 	},
+	ErrInvalidLifecycleWithObjectLock: {
+		Code:           "InvalidLifecycleWithObjectLock",
+		Description:    "The lifecycle configuration is not supported with object locking",
```
```suggestion
		Description:    "The lifecycle configuration with MaxNoncurrentVersions is not supported with object locking",
```

harshavardhana commented on cmd/storage-rest-common.go:
```diff
@@ -18,7 +18,7 @@
 
 const (
-	storageRESTVersion       = "v41" // Optimized DeleteVersions API
+	storageRESTVersion       = "v42" // Added FreeVersions to FileInfoVersions
```
@klauspost looks like its done here

harshavardhana commented on cmd/xl-storage-format-utils.go:
```diff
@@ -63,10 +42,25 @@ func getAllFileInfoVersions(xlMetaBuf []byte, volume, path string) (FileInfoVers
+			versions[n] = fi
+			n++
+		}
+		versions = versions[:n]
+		for _, ver := range versions {
+			ver.NumVersions = n
+		}
```
Should we not add `NumVersions` for XLv1 object as well? @krisis

harshavardhana replied:
@krisis ^^

krisis committed:
Group expiry by days and max-noncurrent-versions under expiryState

krisis committed:
Address review comments

harshavardhana reviewed (changes requested):
One comment everything else LGTM and tested

krisis committed:
Set fi.NumVersions for legacy object (xlv1)

krisis committed:
Add unit test for getFileInfoVersions
